lib/FIR/statements.h

```diff
@@ -581,12 +557,14 @@ class Statement : public SumTypeMixin<ReturnStmt,  //
   }
   std::string dump() const;
 
-  static std::size_t offsetof_impl() {
-    return reinterpret_cast<std::size_t>(&static_cast<Statement *>(nullptr)->u);
-  }
+  // g++/clang++ will optimize this to a simple register copy
   static Statement *From(Stmt_impl *stmt) {
-    return reinterpret_cast<Statement *>(
-        reinterpret_cast<char *>(stmt) - Statement::offsetof_impl());
+    static Statement s{nullptr, UnreachableStmt::Create()};
+    auto *result{reinterpret_cast<Statement *>(reinterpret_cast<char *>(stmt) -
+        (reinterpret_cast<char *>(&s.u) - reinterpret_cast<char *>(&s)))};
+    CHECK(result == reinterpret_cast<Statement *>(stmt) &&
+        "expecting pointers to be equal");
+    return result;
```

Comment:
This still results in accessing a Statement through a pointer that is to a Stmt_impl. I don't think that behavior is defined.

If we need to get from a Stmt_impl to the Statement that contains it, why not use a pointer?

Reply:
The actual point with this is to enforce constraints at the point of FIR construction rather than allow the representation to be constructed incorrectly and then stumble across the problem later.  This is a good thing. Particularly while I am working on the implementation. :)

Specifically for this function, this code is an early detection rather than to just use blind free-casting of the pointer, `reinterpret_cast<Statement*>(p)`. This is also a good thing.

Addressing some possible concerns:

The "guts" of a Statement will never appear on their own without a Statement wrapping them.  I am not worried that they might at some point become separable (as standalone "skin" and "guts"), as that would be a very fundamental and extensive alteration of the structure of the IR.

Concern that the layout of a class's data members in memory will change during execution also seems highly unlikely in practice.  C++ compilers do not (generally, to my knowledge) generate code that will rearrange a class's data members during runtime.

On the standards take: the Statement class is not a POD, so "offsetof" is not well-defined; that is true.  However, Statement, while it is composed of mixins, does not use virtual inheritance, and it is "POD enough" that its data members do not become virtualized indirections, etc.

Conclusion:

I think this can be left for now and handled as a separable TODO issue when FIR becomes more stabilized than it is at present.  Following in the same vein as other parts of this project, I anticipate doing experimentation on how the IR construction constraints can be elevated to the type system at some point.

Hope that helps.

Reply:
Please add a comment to the code with a description of the data structures involved and what logically is happening with From().